go/parquet/compress/compress.go

```diff
@@ -0,0 +1,153 @@
+// limitations under the License.
+
+// Package compress contains the interfaces and implementations for handling compression/decompression
+// of parquet data at the column levels.
+package compress
+
+import (
+	"compress/flate"
+	"io"
+	"io/ioutil"
+
+	"github.com/apache/arrow/go/parquet/internal/gen-go/parquet"
+)
+
+// Compression is an alias to the thrift compression codec enum type for easy use
+type Compression parquet.CompressionCodec
+
+func (c Compression) String() string {
+	return parquet.CompressionCodec(c).String()
+}
+
+// DefaultCompressionLevel will use flate.DefaultCompression since many of the compression libraries
+// use that to denote "use the default".
+const DefaultCompressionLevel = flate.DefaultCompression
+
+// Codecs is a useful struct to provide namespaced enum values to use for specifying the compression type to use
+// which make for easy internal swapping between them and the thrift enum since they are initialized to the same
+// constant values.
+var Codecs = struct {
+	Uncompressed Compression
+	Snappy       Compression
+	Gzip         Compression
+	// LZO is unsupported in this library as I haven't yet found a good implementation of it for golang.
+	Lzo    Compression
+	Brotli Compression
+	// LZ4 unsupported in this library due to problematic issues between the Hadoop LZ4 spec vs regular lz4
+	// see: http://mail-archives.apache.org/mod_mbox/arrow-dev/202007.mbox/%3CCAAri41v24xuA8MGHLDvgSnE+7AAgOhiEukemW_oPNHMvfMmrWw@mail.gmail.com%3E
+	Lz4  Compression
+	Zstd Compression
+}{
+	Uncompressed: Compression(parquet.CompressionCodec_UNCOMPRESSED),
+	Snappy:       Compression(parquet.CompressionCodec_SNAPPY),
+	Gzip:         Compression(parquet.CompressionCodec_GZIP),
+	Lzo:          Compression(parquet.CompressionCodec_LZO),
+	Brotli:       Compression(parquet.CompressionCodec_BROTLI),
+	Lz4:          Compression(parquet.CompressionCodec_LZ4),
+	Zstd:         Compression(parquet.CompressionCodec_ZSTD),
+}
+
+// Codec is an interface which is implemented for each compression type in order to make the interactions easy to
+// implement. Most consumers won't be calling GetCodec directly.
+type Codec interface {
+	// NewReader provides a reader that wraps a stream with compressed data to stream the uncompressed data
+	NewReader(io.Reader) io.ReadCloser
+	// NewWriter provides a wrapper around a write stream to compress data before writing it.
+	NewWriter(io.Writer) io.WriteCloser
+	// NewWriterLevel is like NewWrapper but allows specifying the compression level
+	NewWriterLevel(io.Writer, int) (io.WriteCloser, error)
+	// Encode encodes a block of data given by src and returns the compressed block. dst needs to be either nil
+	// or sized large enough to fit the compressed block (use CompressBound to allocate). dst and src should not
+	// overlap since some of the compression types don't allow it.
+	//
+	// The returned slice *might* be a slice of dst if it was able to fit the whole compressed data in it.
```

Comment:
nit: this seems to disagree with the comment above.  It would imply that a third option is DST will be used if it sized appropriately but otherwise new memory would be allocated.

Reply:
fixed